src/main/java/net/minecraftforge/event/GatherComponentsEvent.java

```diff
@@ -0,0 +1,45 @@
+package net.minecraftforge.event;
+
+import net.minecraft.core.component.DataComponentMap;
+import net.minecraft.core.component.DataComponentType;
+import net.minecraft.world.item.Item;
+import net.minecraftforge.eventbus.api.Event;
+import org.jetbrains.annotations.ApiStatus;
+
+import javax.annotation.Nullable;
+
+/**
+ Where all the events fore Gathering Components will exist.
+ */
+
+public abstract class GatherComponentsEvent extends Event {
+    private final DataComponentMap.Builder components = DataComponentMap.builder();
+
+    public <T> void register(DataComponentType<T> componentType, @Nullable T value) {
+        components.set(componentType, value);
+    }
+
+    @ApiStatus.Internal
+    public DataComponentMap getDataComponentMap() {
+        return components.build();
+    }
+
+
+    public static class GatherItemComponentsEvent extends GatherComponentsEvent {
```

Comment:
Seems simple and functional enough. We've discussed this as a lead in to a gatherCapabilityFactories event system on discord. Where instead of firing the AttachCapabilitiesEvent during every ItemStack instantiation. We gather factories when we gather the components, and then just invoke those factories on each new ItemStack creation. Could even be done lazily. 

Anyways, my only comment on this particular PR is that it could use javadocs, and I vote you rename this Inner class to just `Item`  or move it out to a separate class.